### PR TITLE
Site Editor: Improve loading experience (v1)

### LINKED ISSUE #35503: Refine the Site Editor loading state
Here's the current experience loading the site editor:

https://user-images.githubusercontent.com/846565/136778002-4c8c7264-f37f-4b6f-abe8-524966350380.mp4

A couple of issues to address:

- [ ] The FOWPA (Flash of wp-admin) is awkward. Can we hide it?
- [x] The light grey background feels a bit broken. Can we use the dark grey background from the template editor, and isolated template part editor? https://github.com/WordPress/gutenberg/pull/35520
- [ ] The "Loading..." label in the top bar is easily missed. Something on the canvas would be more obvious.

### PR DESCRIPTION
**Pull request by tyxla**
## What?
This is an experiment - DO NOT MERGE!

This PR aims to experiment with Suspense, specifically with our `useSuspenseSelect()` implementation to load the site editor behind the scenes, and display it once it's finished loading. 

## Why?

We're aiming to find a way to improve the editor loading experience.

As @mtias [says](https://github.com/WordPress/gutenberg/issues/35503#issuecomment-973182125):

> The end goal should be that things load like a "screenshot"; all fully loaded when you start interacting.

See #35503 for more details and motivation.

## How?

We're essentially replacing a bunch of `useSelect()` with `useSuspenseSelect`, adding a `Suspense` boundary with a dummy loading screen, and some artificial delay to prevent from flickering.

This is obviously not landable in this version - we'd likely need a better and more selective approach to which `useSelect` to alter and which not.

## Testing Instructions
* Load the site editor and observe the loading experience. 
* Observe that at the time the fallback placeholder has been hidden, the editor has fully loaded.

## Screenshots or screencast 

https://user-images.githubusercontent.com/8436925/179744565-eaa4deb3-6ffe-4185-b3c0-9c16b01a9973.mov

### CODE REVIEW

**Comment by jameskoster** [comment]
Very excited to see some work on this! :) 

I'm unable to explain why, but I am seeing small flashes of content behind the loader:

https://user-images.githubusercontent.com/846565/179752673-1378299b-5f21-4d07-8457-1c11fb2d0e63.mp4

Let me know how I can help debug that.

It's generally not recommended to use a logo for a loading graphic as it can create a negative association with the brand. I'm not sure how much truth there is to that, but it might be best to use something like the [Spinner](https://wordpress.github.io/gutenberg/?path=/story/components-spinner--custom-size) component for now anyways? Hopefully in the future we can use a determinate progress bar.

**Comment by mtias** [comment]
This is great, thanks for working on it! It'll have a significant impact on the perception of the editor and the fragility waterfalls tend to signal.

**Comment by gziolo** [comment]
@tyxla, impressive work so far. It shows the power of new React APIs ❤️

Have you experimented with giving priorities to suspense calls so we could render some more critical chunks of the page faster?

**Comment by tyxla** [comment]
Thanks for your feedback everyone!

> I'm unable to explain why, but I am seeing small flashes of content behind the loader:

Yes, those are easy to reproduce with CPU throttling. Those flashes will generally occur in periods when there are no async requests in flight, and our suspense boundary "thinks" that loading has finished, thus no longer presents us with the fallback. However, if this happens in the middle of the loading, when a new request is issued shortly, the fallback is displayed again. This is a downside of the lack of control we have over the current mechanism, and demonstrates the need to create a centralized mechanism that allows for better granularity and control (cc @griffbrad as we've been discussing that before). I've also demonstrated this in a simpler and isolated way in #42188. Alternatively, we could hack around this by debouncing the rendering with a small delay, however there's no guarantee it'll cover the flickering, because how short (or how long) a period without requests in the middle of the loading experience can vary greatly. 

> It's generally not recommended to use a logo for a loading graphic as it can create a negative association with the brand. I'm not sure how much truth there is to that, but it might be best to use something like the [Spinner](https://wordpress.github.io/gutenberg/?path=/story/components-spinner--custom-size) component for now anyways?

Of course, this was just for experimentation and demonstration purposes and never meant to land as-is. I've replaced the logo with a spinner component in the meantime if that helps.

> Have you experimented with giving priorities to suspense calls so we could render some more critical chunks of the page faster?

No, I haven't. Actually, I'm not sure this makes a lot of sense for a few reasons:

* As per the discussion in this issue, we'd like to display the page load "as a screenshot", not load section by section. IIUC, that means that we don't load and display certain parts of the page faster; we display it all at the same time, when everything has loaded and is ready for display.
* How would you decide what to prioritize? To ensure everything will have loaded when finally displaying the editor, we need to asynchronize all requests. At the same time, we don't know what data is going to be necessary, because some of the components that will issue requests haven't still been mounted. Also the data is dynamic, and may (or not) contain various pieces, where priority may vary between different themes and templates. Does that infer the creation of a mechanism to point out the priority? Maybe, but it would add a lot of complexity and I'd really like us to be able to improve the loading experience without requiring extra APIs to be used.

Of course, I might be misunderstanding, so let me know if that's the case.

**Comment by mtias** [comment]
> It's generally not recommended to use a logo for a loading graphic as it can create a negative association with the brand.

@jameskoster I think this is kind of nonsense :D 

> As per the discussion in this issue, we'd like to display the page load "as a screenshot", not load section by section. IIUC, that means that we don't load and display certain parts of the page faster; we display it all at the same time, when everything has loaded and is ready for display.

Correct. We might want to identify some structural parts like editor header and navigation bits (W menu) first and display those as soon as ready so you can go elsewhere and not wait for everything to load before navigating.

**Comment by tyxla** [comment]
> We might want to identify some structural parts like editor header and navigation bits (W menu) first and display those as soon as ready so you can go elsewhere and not wait for everything to load before navigating.

Oh, interesting. I'd love to see some sort of a raw preview or demo of how you envision this working so I could tinker with it.

**Comment by jameskoster** [comment]
> @jameskoster I think this is kind of nonsense :D

Maybe :D 

> I'd love to see some sort of a raw preview or demo

A very rough idea:

load

Note how the W button finishes loading first, then the top bar, then the canvas. 

I do wonder about detaching the top bar / canvas though. What happens if you open the Inspector before the canvas is loaded? Or try to add a block via the Inserter? I'm not sure there's any real benefit to loading those tools before the canvas. It might be better to simply load those together. An alternative:

load2

**Comment by tyxla** [comment]
Thanks for the demo! 

At a first glance, this would be a bit troublesome with the current interface skeleton structure. Let me clarify.

Currently, we have this rough structure for the site editor:

- Site Editor app
	- Suspense boundary
		- Editor
			- Interface Skeleton
				- Header
				- Content
				- Sidebar


To achieve what's suggested on the video, we'd have:

- Site Editor app
	- Editor
		- Interface Skeleton
			- Header
				- Suspense boundary
			- Content
				- Suspense boundary
			- Sidebar
				- Suspense boundary

However, for this to work, we need to have no async suspense select calls outside of suspense boundaries. This is tough, because the Editor and the Header, for example, will share some data dependencies, and we can't (and shouldn't) query those data dependencies both ways. 

To be able to do that, we'd need to structurally pull out the header from the interface skeleton, so it would be outside of the Editor, in order for the Editor and the Header to have separate suspense boundaries.

We should be able to get around this with `useTransition()`, however, this feature is available from React 18, and AFAIK this requires us to also upgrade to React Native 0.69 (https://reactnative.dev/blog/2022/06/21/version-069) but we're still stuck on an older version.

That being said, I wonder what still prevents us from upgrading to React 18 and if we can push that forward. cc @gziolo

**Comment by Mamaduka** [comment]
> However, for this to work, we need to have no async suspense select calls outside of suspense boundaries.

We also probably won't need to use suspense select everywhere. Maybe start with selectors that fetch entities from the server.

**Comment by mtias** [comment]
@tyxla @jameskoster we don't need anything too complicated at first here. Rendering a very basic skeleton (without pulsing elements) just outlining the header and site icon slot is fine. More important to get the foundation in so we can refine.

**Comment by jameskoster** [comment]
Here's this PR as things stand:

loader

It's already in a decent spot imo, and potentially mergeable assuming subsequent refinements. If it's easy to try, we could use a slightly larger spinner, and print a very simple indication of where the site logo / top bar will appear:

Screenshot 2022-08-17 at 09 48 27

**Comment by mtias** [comment]
Yep, that last one is what I had in mind

**Comment by jameskoster** [comment]
One issue to fix: The loading state should (for now) only be displayed when you first open the Site Editor. At the moment I noticed it pops up when you perform actions like:

* Click 'Replace' in the Template Part toolbar
* Open Global Styles
* Switch to the Templates list (shouldn't this be included in the initial load?)

**Comment by tyxla** [comment]
Thanks for the feedback everyone!

I've done some changes and here's how it looks now:

https://user-images.githubusercontent.com/8436925/185119864-4cf1474e-2816-4073-8b9c-2c0d3cc6c54c.mov

Basically, the header is now outside and since it generally loads quickly and doesn't have a ton of dependencies, I don't see a reason to add suspense to it at this time. We could, of course, do it if we find a good use case that makes it load slowly.

Beyond that, I've moved the boundary for the content and since will take more time to load, we'll see the spinner. I've increased the size of that a bit, just to demonstrate how things look that way. 

Let me know what you think!

> The loading state should (for now) only be displayed when you first open the Site Editor

I'd be happy to tinker with that, but was curious - what happens when we start interacting with the app and new data gets requested? Is it going to work just like it did before - with placeholders and spinners all over the place?

**Comment by jameskoster** [comment]
It's not a strong opinion, but it's a little strange to have access to things like the Inserter before the document is visible. It might be worth trying a version where the icons/buttons in the top bar are hidden until the document loads.

If you click the W button during the loading state, the menu isn't visible:

Screenshot 2022-08-17 at 15 40 19

**Comment by mtias** [comment]
Can we get this one through the finish line?

**Comment by tyxla** [comment]
I plan to focus on addressing the feedback and getting it to a more shippable state next week.

**Comment by jsnajdr** [comment]
I've been playing with this PR, trying to understand what exactly is going on when the Site Editor loads. And discovered some inefficiencies. I noticed that during the editor load, the `PostTemplateEdit` component rerenders 86 times. That's a lot -- we don't have that many moving parts and network loads that would justify such a high number.

While investigating this, I found the [`useTaxonomiesInfo` hook](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/query/edit/inspector-controls/taxonomy-controls.js#L36-L55). This hook is inefficient and causes many wasted rerenders. It's because its `useSelect` callback returns a different value (`taxonomiesInfo`) each time it's called, even though the taxonomies data in the underlying store (`coreStore`) haven't changed at all. That's because the return value is a result of `.map`.

The result is that every time anything in the `coreStore` changes, all components calling `useTaxonomiesInfo` will be forced to rerender.

Moving the `.map` call outside `useSelect` is not easy because the `.map` callback uses `getEntityRecords` selector for each taxonomy type. A more complex refactoring will be needed.

One further inefficiency is that most of the time we don't use the `useTaxonomiesInfo` result at all, even though we load it. By default, the Query Loop block has the "Inherit query from template" option turned on, meaning that it doesn't have a custom query to edit. It's only when we're editing a custom query for the block when the taxonomies info is actually used.

This functionality was added by @ntsekouras in #38063.

I suspect the Site Editor load will contain further inefficiencies like this, to be discovered after we fix this one.

**Comment by jsnajdr** [comment]
One area where the Suspense spinner adds some new dynamic is block visibility. There is an intersection observer observing block DOM elements and setting visibility flags in `blockEditorStore`. That info is then used to switch blocks between sync and async data mode as they scroll in and out of view.

Now visibility is being set also when the spinner is being displayed. The block editor UI is actually in DOM all the time, just hidden with `style="display: none !important"`. There is not just the spinner. As new blocks are being progressively rendered, they are detected as hidden by the intersection observer, and data state updates are triggered.

I don't see any that would be buggy or harmful, just some not exactly expected state activity (and associated rerenders) that's redundant in this context.

**Comment by mtias** [comment]
I'm curious if `content-visibility` would be a better native mechanism for coming-into-view at some point.

**Comment by jsnajdr** [comment]
It's been already discussed that there is often a flash of content while loading, where the spinner disappears for a while in the middle of the load. I tried to find out how exactly it happens and discovered that it's a combination of classic non-suspense `useSelect` and the new `useSuspenseSelect`. Imagine three nested components where the first and third load data with Suspense, while the second, middle one loads traditionally:
```js
function First() {
  const { firstData } = useSuspenseSelect( select => { ... } );
  return <Second data={ firstData } />;
}

function Second() {
  const { secondData, hasResolved } = useSelect( select => { ... } );
  if ( ! hasResolved ) {
    return <LocalSpinner />;
  }
  return <Third data={ secondData } />;
}

function Third() {
  const { thirdData } = useSuspenseSelect( select => { ... } );
  return <div>{ thirdData.text }</div>;
}
```
Here we have a cascade of three data loads, where each waits one for the previous one to load before it starts. Because a component is not even rendered and mounted before the parent's load finishes. It's an anti-pattern but anyway.

The waiting is implemented either by Suspense throwing and suspending, or by rendering conditionally only after `hasResolved` becomes `true`. If all three components used Suspense, the rendering would gradually progress forward, and suspending on each unfinished load. The global spinner would be rendered the entire time.

But here, after the first load finishes, Suspense no longer suspends and renders a "complete" UI with a local spinner inside `<Second />`. Then `<Third />` suspends again and you see the global spinner again. That's exactly the flash.

In practice, the "Page List" block that's part of the default Twenty Twenty-Two template behaves exactly this way:

Screenshot 2022-09-12 at 16 17 37

It loads the page list without Suspense, with its `usePageData` hook, and then the inner `PageList` components loads stuff with Suspense, in its `useFrontPageId` hook. Source: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/page-list/edit.js

That's how you get the content flash when displaying a page with that block. If `usePageData` used Suspense, too, the loading experience would be smooth.

**Comment by ockham** [comment]
@tyxla @jsnajdr Thanks a lot for working on this! As @mtias said, we'd love to include it in WP 6.1. In order for that to happen, we should land this PR before Feature Freeze (September 20) -- ideally with a bit of leeway to cherry-pick and publish new npm packages for use by Core.

Do you think that'll be enough time? (Is there any chance you could prioritize it?) Thanks in advance! 😊🙌

**Comment by tyxla** [comment]
@mtias @ockham I'm prioritizing this till the end of this week.

**Comment by tyxla** [comment]
I've spent some time with this today, rebasing it and working on the feedback here. Here's a preview:


https://user-images.githubusercontent.com/8436925/190165333-5b393c44-a819-48e5-a80d-9e45774102e4.mov



Since the last version, now we've addressed all remaining `useSelect()` instances in the block library to use `useSuspenseSelect`, which @jsnajdr suggested in an earlier comment as it was causing unexpected flickering. For what it's worth, some of the `useSelect` usages were introduced after the PR was opened, which is why there were so many of them remaining.

I've also improved the placeholder to use a `<Modal />` which allows us to occupy the entire screen with the overlay in order to disallow header interactions - something that came up in feedback by @jameskoster earlier. I've tinkered with the modal styles a bit - removing the overlay background but keeping it to disallow header interactions. I've added a white background to the modal frame and updated it to extend to the entire screen.

Another feature I've tinkered with is displaying the loading placeholder only on initial loading, as suggested by @jameskoster and @mtias. You can see in the preview that the image and page list no longer cause the loading screen to appear. Admittedly, it's in a hacky state, because with `Suspense` we don't have the luxury to have it display only on the first load; Suspense doesn't care whether it's the first or last load. So this requires some hacks, including runtime interchanging of `useSuspenseSelect` to work like `useSelect` after loading, and some artificial delay after loading to conceal the rerender flicker of elements that occurs when we re-render the editor without the suspense boundary. 

Something I don't particularly like is that because we use the same blocks in all editors, that means that we need a suspense boundary in all editors, so I've added one to the post editor as well. I'm not particularly happy with this, and one way to get around it might be to trick `useSuspenseSelect()` to work like `useSuspense()` for the other editors, as we did after the loading screen gets away. 

One downside with the way things are right now is that React Native doesn't have a `Modal` at all, so we have to reconsider the native implementation for the loading screen. We might want to keep it simple there. For now, I've reverted the suspense changes there, although it would be nice to explore a better loading experience there as well.

Of course, this needs some more love, too. We've added the loading screen context in the data package for the purpose of the experiment, but there will surely be a better place for it. There might be more flickering here and there depending on the theme, templates and blocks used, and unfortunately, there aren't many ways to get around it other than adding artificial timeouts, because right now we don't have a whole lot of control and granularity with Suspense and how it manages the placeholder. We might want to extend `Suspense` to add some additional control over how and when we display the fallback and the children components. 

To summarize, this is still an experiment and is not production-ready, and I doubt we can polish it well soon enough without breaking stuff, so it might be best to postpone it.

I'd love to get some feedback, and I'm happy for y'all to chime in with suggestions and updates to the PR directly. Thanks!

**Comment by jsnajdr** [comment]
One major downside is that all blocks, even the 3rd party ones that we don't control, need to use Suspense consistently. Otherwise, when classic and Suspense data loading is mixed together, the loading indicator will flicker.

Maybe we don't need to use Suspense at all. When the editor UI is suspended, it means that the DOM is being incrementally built inside a hidden element (`display: none !important`), and the fallback spinner is displayed. We could approximately implement the same thing without Suspense: while any data are loading, hide the UI with `display` or `content-visibility`, and reveal it only after everything is done.

The biggest difference is running effects. Suspense won't run any effects while the UI is suspended -- it will only build the DOM. A normal hidden `div`, however, won't prevent effects from running. That could have some unwanted consequences for, e.g., effects that measure DOM dimensions, which are all zero.

**Comment by ndiego** [comment]
Hello all, can I get a status update on this one? Given that 6.1 Beta 1 has been released, can I remove this PR from the 6.1 Project Board and remove the Backport Label? @mtias @tyxla

**Comment by jsnajdr** [comment]
@ndiego This work is unfortunately still very experimental and not likely to be in 6.1. You can safely remove it from any 6.1 project tracking stuff.

**Comment by ndiego** [comment]
Thanks for the clarification @jsnajdr, removing the backport label.

**Comment by tyxla** [comment]
> One major downside is that all blocks, even the 3rd party ones that we don't control, need to use Suspense consistently. Otherwise, when classic and Suspense data loading is mixed together, the loading indicator will flicker.

Indeed. One way around this would be having a mechanism to *always* decide on runtime whether to use `useSelect` or `useSuspenseSelect`, and going with that for all usages that use `useSelect` as well. That way plugins would also benefit from that improvement. I still think it's worth experimenting with that. It might solve some of the problems created by the current approach we're exploring here.

**Comment by youknowriad** [comment]
> Another feature I've tinkered with is displaying the loading placeholder only on initial loading, as suggested by @jameskoster and @mtias. You can see in the preview that the image and page list no longer cause the loading screen to appear. Admittedly, it's in a hacky state, because with Suspense we don't have the luxury to have it display only on the first load; Suspense doesn't care whether it's the first or last load. So this requires some hacks, including runtime interchanging of useSuspenseSelect to work like useSelect after loading, and some artificial delay after loading to conceal the rerender flicker of elements that occurs when we re-render the editor without the suspense boundary.

After reading a bit on the internet (as this seemed like a common issue people could face), it seems the solution for folks here is to actually use "startTransition" to avoid showing the fallbacks. https://reactjs.org/docs/react-api.html#starttransition


So is it possible to have this:

 - `useSuspenseSelect` only for the important blocks we want to include in the loading fallback: template parts, navigation block as a start maybe, potentially images. (I don't think we necessarily need all blocks and third-party blocks to suspend rendering by default)
 - wrap all our editors with `Suspense` (to avoid breakage in the editors)
 - Add `startTransition` around the changes that we don't want to impact the loading state with (like `setAttributes` maybe)

What would be the issues with such approach?

**Comment by youknowriad** [comment]
(I guess we need to wait for React 18 for startTransition to be available)

**Comment by tyxla** [comment]
I've been hoping to use `startTransition` and `useTransition` here, and that's been one of my motivators to push forward to migrate all the remaining enzyme tests. The PR to remove `enzyme` is up now, by the way: 

* https://github.com/WordPress/gutenberg/pull/44494

Which means the upgrade to React 18 will be unblocked very soon and we can reattempt it.

**Comment by mtias** [comment]
Where are we on the update to 18  blocker?

**Comment by Mamaduka** [comment]
Here's the latest PR by @jsnajdr - #45235.

**Comment by tyxla** [comment]
Now that we're ready with upgrading the dependencies in #45235, we're working to enable concurrent rendering features in #46467, which is the last blocker before we're able to experiment with `useSyncExternalStore` and `startTransition` / `useTransition`.

**Comment by jsnajdr** [comment]
> Add `startTransition` around the changes that we don't want to impact the loading state with (like `setAttributes` maybe)

`startTransition` doesn't affect updates done by `useSyncExternalStore`. These updates will always be sync (`queueMicrotask`). In code like:
```js
const { setAttributes } = useDispatch();
startTransition( () => {
  setAttributes( newAttributes );
);
```
the `startTransition` code will be a noop: it will just synchronously execute the callback and won't do anything else.

I don't fully understand the problem we hope to solve with `startTransition` -- what do we want to delay and when? Then we can look for solutions.

**Comment by tyxla** [comment]
Some context on what @jsnajdr talks about above can be found in this discussion: https://github.com/reactwg/react-18/discussions/86?sort=top

Particularly this:

> Updates triggered by a store change will always be synchronous, even when wrapped in `startTransition`

I don't think we considered experimenting with `startTransition` together with `useSyncExternalStore` in the first place. In any case, now that we use `useSyncExternalStore` in both `useSelect` and `useSuspenseSelect`, we'll need to drop the idea of using the transition API for managing the loading state, unless I'm misunderstanding something. 

I'm starting to doubt if with suspense we will be able to solve this problem altogether. The issue there is all the third-party blocks and plugins which will not be suspending rendering in their components. We can't realistically ask for every plugin to suspend, therefore, we might need an alternative implementation.

This brings me back to an alternative solution we discussed earlier in the summer with @jsnajdr - basically implementing a custom wrapper system that relies on the status of the resolvers we're interested in and based on that, rendering either the placeholders or the actual component.

@jsnajdr @youknowriad do you have any additional thoughts or ideas on moving forward with this one?

**Comment by youknowriad** [comment]
> I'm starting to doubt if with suspense we will be able to solve this problem altogether. The issue there is all the third-party blocks and plugins which will not be suspending rendering in their components. We can't realistically ask for every plugin to suspend, therefore, we might need an alternative implementation.

Personally, I don't see this as a blocker at all. The important thing is that we get "some loaders" absorbed by the suspense provider, not all of them. Blocks like reusable blocks, template parts and navigation blocks are the most important one here, maybe images as well but if there's some blocks including third-party blocks that load asynchronously (have their own loaders), that's totally fine.

**Comment by mtias** [comment]
I agree — we don't need to absorb everything for this to succeed in addressing the main problem. It's ok to have a few things here and there waterfall loading; it's not ok for the main pieces of the interface and fundamental building blocks to load uncoordinated. Let's please start small and make the initial experience better than what it is today while we figure out the more difficult parts and how they integrate with the block API for other authors to take advantage of.

**Comment by tyxla** [comment]
Thanks for the feedback, everyone!

I'm working on another experiment today - similar to what we had here, but a bit simpler. My idea is to still use suspense but to try declaring dependencies per the suspense boundary. That will not only give us the required granularity to add multiple suspense boundaries (header, content, footer) but will also allow us to have the flexibility and add/edit data dependencies for each suspense boundary as needed. 

PR coming soon.

**Comment by tyxla** [comment]
Here's the PR of this other experiment I was referencing above: https://github.com/WordPress/gutenberg/pull/47612

**Comment by tyxla** [comment]
I've started working on a v3, which I prefer the most from the 3 approaches: #50222.

**Comment by tyxla** [comment]
Since #50222 is much more reliable and flexible, I'm closing this one in its favor.
